lib/photo_exporter.rb

```diff
@@ -0,0 +1,43 @@
+class PhotoExporter
+  attr_reader :user
+
+  def initialize(user)
+    @user = user
+  end
+
+  def perform
+    temp_zip = Tempfile.new([user.username, "_photos.zip"])
+    begin
+      Zip::OutputStream.open(temp_zip.path) do |zip_output_stream|
+        user.photos.each do |photo|
+          export_photo(zip_output_stream, photo)
+        end
+      end
+    ensure
+      temp_zip.close
+    end
+
+    update_exported_photos_at(temp_zip)
+  end
+
+  private
+
+  def export_photo(zip_output_stream, photo)
+    photo_file = photo.unprocessed_image.file
+    if photo_file
+      photo_data = photo_file.read
+      zip_output_stream.put_next_entry(photo.remote_photo_name)
+      zip_output_stream.print(photo_data)
+    else
+      user.logger.info "Export photos error: No file for #{photo.remote_photo_name} not found"
+    end
+  rescue Errno::ENOENT
+    user.logger.info "Export photos error: #{photo.unprocessed_image.file.path} not found"
+  end
+
+  def update_exported_photos_at(temp_zip)
+    user.update exported_photos_file: temp_zip, exported_photos_at: Time.zone.now if temp_zip.present?
+    user.restore_attributes if user.invalid? || temp_zip.present?
+    user.update exporting_photos: false
+  end
```

Comment:
So I guess `ensure` was lost unintentionally.

I don't follow it either, I guess it was quite wrong. The thing is that `temp_zip.present?` returns true everywhere here even if `temp_zip.close` or `temp_zip.unlink` called. Therefore `user.restore_attributes` is called everytime. However for the successful save case the  `user.restore_attributes` is no-ops because it restores attributes for dirty model, while `user.update` saves model and therefore makes the model instance clean. So I think we can safely remove `|| temp_zip.present?` condition.